spec/Section 2 -- Language.md

```diff
@@ -938,3 +943,292 @@ and operations.
+
+```graphql
+scalar DateTime
+```
+
+
+#### Object
+
+ObjectTypeDefinition : type Name ImplementsInterfaces? Directives? { FieldDefinition+ }
+
+ImplementsInterfaces : implements NamedType+
+
+FieldDefinition : Name ArgumentsDefinition? : Type Directives?
+
+ArgumentsDefinition : ( InputValueDefinition+ )
+
+InputValueDefinition : Name : Type DefaultValue? Directives?
+
+Object types represent a list of named fields, each of which yield a value of a
+specific type. Each field itself may accept a list of named arguments.
+
+Objects may implement Interface types. When implementing an Interface type, the
+Object type must supply all fields defined by the Interface.
+
+In this example, a Object type called `TodoItem` is defined:
+
+```graphql
+type TodoItem implements Node {
+  id: ID
+  title: String
+  isCompleted: Boolean
+}
+```
+
+#### Interface
+
+InterfaceTypeDefinition : interface Name Directives? { FieldDefinition+ }
+
+Interface types, similarly to Object types represent a list of named fields.
+Interface types are used as the type of a field when one of many possible Object
+types may yielded during execution, but some fields are guaranteed. An Object
+type is a possible type of an Interface when it both declares that it implements
+that Interface as well as includes all defined fields.
+
+In this example, an Interface type called `Node` is defined:
+
+```graphql
+interface Node {
+  id: ID
+}
+```
+
+#### Union
+
+UnionTypeDefinition : union Name Directives? = UnionMembers
+
+UnionMembers :
+  - NamedType
+  - UnionMembers | NamedType
+
+Union types represent a list of named Object types. Union types are used as the
+type of a field when one of many possible Object types may yielded during
+execution, and no fields are guaranteed. An Object type is a possible type of a
+Union when it is declared by the Union.
+
+In this example, a Union type called `Actor` is defined:
+
+```graphql
+union Actor = User | Business
+```
+
+
+#### Enum
+
+EnumTypeDefinition : enum Name Directives? { EnumValueDefinition+ }
+
+EnumValueDefinition : EnumValue Directives?
+
+EnumValue : Name
+
+Enum types, like Scalar types, also represent leaf values in a GraphQL type
+system. However Enum types describe the set of legal values.
+
+In this example, an Enum type called `Direction` is defined:
+
+```graphql
+enum Direction {
+  NORTH
+  EAST
+  SOUTH
+  WEST
+}
+```
+
+#### Input Object
+
+InputObjectTypeDefinition : input Name Directives? { InputValueDefinition+ }
+
+Input Object types represent complex input values which may be provided as an
+field argument. Input Object types cannot be the return type of an Object or
+Interface field.
+
+In this example, an Input Object called `Point2D` is defined:
+
+```graphql
+input Point2D {
+  x: Float
+  y: Float
+}
+```
+
+### Type Extension
+
+TypeExtensionDefinition : extend ObjectTypeDefinition
+
+Type extensions may be used by client-side tools in order to represent a
```

Comment:
This could also be used on the server if you are using the IDL to define your schema, but want to split one type across multiple files. In particular, this would be common for `Query` or `Mutation`, where you might want to do:

```graphql
type Todo {
  text: String!
}

extend type Query {
  allTodos: [ Todo ]
}
```

In the super ideal case, `buildASTSchema` would take this into account and merge fields.

Relevant issue on graphql-tools: https://github.com/apollostack/graphql-tools/issues/182